tests/jsunit/xml_test.js

```diff
@@ -131,31 +156,123 @@ function test_domToText() {
   try {
     var dom = Blockly.Xml.textToDom(
-        '<xml xmlns="http://www.w3.org/1999/xhtml">' +
-        '  <block type="test_block" inline="true" x="21" y="23">' +
+        '<xml>' +
+        '  <variables>' +
+        '    <variable type="type1" id="id1">name1</variable>' +
+        '    <variable type="type2" id="id2">name2</variable>' +
+        '    <variable type="" id="id3">name3</variable>' +
+        '  </variables>' +
+        '  <block type="field_variable_test_block">' +
+        '    <field name="VAR" id="id3" variabletype="">name3</field>' +
         '  </block>' +
         '</xml>');
     Blockly.Xml.domToWorkspace(dom, workspace);
     assertEquals('Block count', 1, workspace.getAllBlocks().length);
+    xmlTest_checkVariableValues('name1', 'type1', 'id1');
+    xmlTest_checkVariableValues('name2', 'type2', 'id2');
+    xmlTest_checkVariableValues('name3', '', 'id3');
+  } finally {
+    delete Blockly.Blocks.test_block;
+    xmlTest_tearDownWithMockBlocks();
+  }
+}
+
+function test_domToWorkspace_VariablesAtTop_DuplicateVariablesTag() {
+  // Expect thrown Error because of duplicate 'variables' tag
+  xmlTest_setUpWithMockBlocks();
+  try {
+    var dom = Blockly.Xml.textToDom(
+        '<xml>' +
+        '  <variables>' +
+        '  </variables>' +
+        '  <variables>' +
+        '  </variables>' +
+        '</xml>');
+    Blockly.Xml.domToWorkspace(dom, workspace);
+    fail();
+  }
+  catch (e) {
+    // expected
   } finally {
     delete Blockly.Blocks.test_block;
+    xmlTest_tearDownWithMockBlocks();
+  }
+}
 
+function test_domToWorkspace_VariablesAtTop_MissingType() {
+  // Expect thrown error when a variable tag is missing the type attribute.
+  workspace = new Blockly.Workspace();
+  try {
+    var dom = Blockly.Xml.textToDom(
+        '<xml>' +
+        '  <variables>' +
+        '    <variable id="id1">name1</variable>' +
+        '  </variables>' +
+        '  <block type="field_variable_test_block">' +
+        '    <field name="VAR" id="id1" variabletype="">name3</field>' +
+        '  </block>' +
+        '</xml>');
+    Blockly.Xml.domToWorkspace(dom, workspace);
+    fail();
+  } catch (e) {
+    // expected
+  } finally {
     workspace.dispose();
   }
 }
 
+function test_domToWorkspace_VariablesAtTop_MismatchBlockType() {
+  // Expect thrown error when the serialized type of a variable does not match
+  // the type of a variable field that references it.
+  xmlTest_setUpWithMockBlocks();
+  try {
+    var dom = Blockly.Xml.textToDom(
+        '<xml>' +
+        '  <variables>' +
+        '    <variable type="type1" id="id1">name1</variable>' +
+        '  </variables>' +
+        '  <block type="field_variable_test_block">' +
+        '    <field name="VAR" id="id1" variabletype="">name1</field>' +
+        '  </block>' +
+        '</xml>');
+    Blockly.Xml.domToWorkspace(dom, workspace);
+    fail();
+  } catch (e) {
+    // expected
+  } finally {
+    delete Blockly.Blocks.test_block;
```

Comment:
I think you mean to delete `Blockly.Blocks.field_variable_test_block`, but you can do it in `xmlTest_tearDownWithMockBlocks` since you're adding it in `xmlTest_setUpWithMockBlocks`